pilot/pkg/networking/v1alpha3/clusters.go

```diff
@@ -0,0 +1,210 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha3
+
+import (
+	"fmt"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+
+	v2_cluster "github.com/envoyproxy/go-control-plane/envoy/api/v2/cluster"
+	"github.com/gogo/protobuf/types"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+// BuildClusters returns the list of clusters for the given proxy. This is the CDS output
+func BuildClusters(env model.Environment, node model.Proxy) []*v2.Cluster {
+	var clusters []*v2.Cluster
+
+	// TODO node type router
+
+	// node type Sidecar
+	services, err := env.Services()
+	if err != nil {
+		log.Errorf("Failed for retrieve services: %v", err)
+		return clusters
+	}
+
+
+	// build Outbound Clusters
+	for _, service := range services {
+		config := env.IstioConfigStore.DestinationRule(service.Hostname, "")
+		if config != nil {
+			destinationRule := config.Spec.(*networking.DestinationRule)
+			for _, svcPort := range service.Ports {
+
+				var hosts []*core.Address
+				if service.Resolution == model.DNSLB {
+					// FIXME port name not required if only one port
+					instances, err := env.ServiceDiscovery.Instances(service.Hostname, []string{svcPort.Name}, nil)
+					if err != nil {
+						log.Errorf("failed to retrieve instances for %s: %v", service.Hostname, err)
+					} else {
+						for _, instance := range instances {
+							hosts = append(hosts, &core.Address{Address: &core.SocketAddress{
+								Address: instance.Endpoint.Address,
+								Protocol: core.TCP,
+								PortSpecifier: &core.SocketAddress_PortValue{
+									PortValue: uint32(instance.Endpoint.Port),
+								},
+							}})
+						}
+					}
+				}
+
+				// create default cluster
+				cluster := &v2.Cluster{
+					Name: buildClusterName(service.Hostname, "", svcPort.Name),
+					Type: convertResolution(service.Resolution),
+					Hosts: hosts,
+				}
+				applyTrafficPolicy(cluster, destinationRule.TrafficPolicy)
+				clusters = append(clusters, cluster)
+
+				for _, subset := range destinationRule.Subsets {
+					cluster := &v2.Cluster{
+						Name: buildClusterName(service.Hostname, subset.Name, svcPort.Name),
+						Type: convertResolution(service.Resolution),
+						Hosts: hosts,
+					}
+					applyTrafficPolicy(cluster, destinationRule.TrafficPolicy)
+					applyTrafficPolicy(cluster, subset.TrafficPolicy)
+
+					clusters = append(clusters, cluster)
+				}
+
+				// TODO HTTP2 cluster feature
+
+			}
+		}
+	}
+
+	// TODO build inbound clusters
+
+	// TODO build TCP clusters in/out
+
+	// TODO add one original dest cluster
+
+	return clusters
+}
+
+// TODO port name is not mandatory if only one port defined
+// TODO subset name can be empty
+func buildClusterName(hostname, subsetName, portName string) string {
```

Comment:
For 0.8, please use the same name that v1 is using. 

It is very possible RDS v2 will not be ready, and it will keep the option of mix/match. There is no immediate benefit of renaming cluster names at this point - we can do it after 0.8.

Reply:
Please do not carry over the cluster naming. It beats the whole purpose of this PR. I do not care if RDSv2 is ready. Kuat can fix it in no time. Using the old style cluster naming brings in the need to do label selectors, etc. and complicates the CDS code.

In fact, if we form the cluster name in this way, RDS code will not have to touch destination rules or even bother about CDS stuff. It can form the cluster name in exactly the same manner.